Mixed precision Grappler optimizer

MattConley opened the pull request:
Adds an opt-in Grappler pass to convert parts of a graph from float32 to float16.

Thanks to @minminsun, @Dido0o0, and @yangjunpro of Alibaba for their useful collaboration and feedback on this work.

Attention @azaks2 and @reedwm, and FYI @jlebar.

MattConley committed:
Add mixed precision Grappler optimizer

 - Adds an opt-in Grappler pass to convert parts of a graph from float32 to float16

Co-authored-by: Ben Barsdell <bbarsdell@nvidia.com>
Co-authored-by: Carl Case <carlc@nvidia.com>
Co-authored-by: Trent Lo <trentl@nvidia.com>
Co-authored-by: Nathan Luehr <nluehr@nvidia.com>

jlebar commented:
@rthadur, since this doesn't touch XLA, I don't think I am the right reviewer for this PR.

MattConley committed:
Remove ENV variable enabling grappler pass

- Keeping the wrapper function for use when loss scaling interaction
  is finalized

penpornk commented:
@rthadur I think @reedwm and @rmlarsen are the more appropriate people for this PR, so I'm removing myself as well.

byronyi commented:
This PR is huge and proposes nontrivial changes. It would be helpful to draft a design document or an RFC to facilitate the review process.

benbarsdell commented:
We're looking into the compilation errors now.

benbarsdell committed:
Enable GOOGLE_CUDA guard around cuda include

benbarsdell committed:
Add checks when using absl::optional return values

- This is primarily to WAR what appears to be a compiler bug in
  GCC <= 4.9.0, which claims that there is an ambiguity between const T&
  and const T&& overloads of the absl::optional::value() method when
  the object is an rvalue reference. Later versions of GCC (and Clang
  generally) do not appear to have any problem with this construct.

benbarsdell committed:
Fix formatting of auto_mixed_precision_lists.h

MattConley committed:
Fix python lint for auto_mixed_precion_test.py

MattConley committed:
Fix future imports in auto_mixed_precision_test.py

MattConley committed:
Fix order of meta_optimizer build dependencies

reedwm commented on tensorflow/python/grappler/auto_mixed_precision_test.py:
```diff
@@ -0,0 +1,555 @@
+    weights = _weight(input_tensor.get_shape().as_list())
+    return math_ops.matmul(input_tensor, weights, name=name)
+  if color == 'g': # Gray node
+    return math_ops.sqrt(input_tensor, name=name)
```
Having sqrt and log is fine, but will cause NaNs on negative numbers, and in tests, you assert variables are all close. NaN is close to NaN but that's not a useful check, so change these to values that do not cause NaNs, or add a TODO.

benbarsdell replied:
You're right, these tests were producing NaNs and silently failing.
I've changed them to NaN-proof ops.

reedwm commented on tensorflow/python/grappler/auto_mixed_precision_test.py:
```diff
@@ -0,0 +1,555 @@
+class AutoMixedPrecisionTest(test.TestCase):
+  """Tests the Grappler auto mixed precision optimizer."""
+  MIN_GPU_ARCH = (7, 0)
+
```
Add a TODO to add eager tests with a tf.function.

One potential issue is that with V2 tf.cond (and i think tf.while_loop), they use tf.functions for the predicate and bodies instead of Switch and Merge ops. I think in eager, V2 cond is done by default. This will force the boundaries of tf.conds and tf.while_loops to be in float32.

benbarsdell replied:
I've added a TODO.

reedwm commented on tensorflow/python/grappler/auto_mixed_precision_test.py:
```diff
@@ -0,0 +1,555 @@
+      self._assert_output_fp16(node_map, 'while/Tanh_1')
+      self.assertAllClose(output_val_ref, output_val, atol=1e-3, rtol=1e-3)
+
+  def test_propagation_through_intertwined_loop_1(self):
```
I think having a few loop tests that are intertwined would be nice. We have one but it is very simple. I am worried someone will break these tests, but not any others, and be confused about why the expected dtypes are these values. Adding a TODO is fine.

benbarsdell replied:
Given your comment about intertwined loops and the fact that the previous tests were mostly producing NaNs, I've replaced them with a set of simple loop tests.

reedwm commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -0,0 +1,1762 @@
+
+// This is just like GraphTopologyView but with (NodeDef, TypeAttrId) pairs as
+// the vertices instead of just NodeDef.
+class GraphTypeTopologyView {
```
I think using a GraphTopologyView would suffice. I think including the types in the graph could hide certain relevant connections.

E.g., if we have an op with two inputs of type T and U respectively, and two outputs of type T and U. With GraphTypeTopologyView, the op would be represented by two unconnected nodes. But, U could affect the numerical stability of the T output, so there perhaps should be a connection.

In practice, there is no such op. But in practice, I also think a GraphTopologyView would behave similarly to a GraphTypeTopologyView and not impact correctness or performance. Duplicating code for a GraphTypeTopologyView is not great.

For now, the current approach is fine, bu we can discuss about changing this after this PR is submitted.

benbarsdell replied:
I've added a TODO.

rmlarsen replied:
I would prefer that we do not check all this duplicated code that is to some extent copy-and-pasted from existing Grappler code.

reedwm replied:
While I agree the duplicated code is not great, I think we should leave it as a TODO and try changing this in a future PR. Reducing this duplication will result in fundamentally changing the algorithm used to color nodes and insert casts, and this new hypothetical algorithm is currently untested and may not work. Also, we are trying to get this PR in for the TF 1.14 branch, so that users can start using mixed precision as soon as possible.

rmlarsen replied:
Sounds good to me.

reedwm reviewed (changes requested):
Thank you for the PR! This is looking very good.

I have a lot of comments but most are nits. Since we want to get this in fairly quickly, feel free to add TODOs for any comments that are difficult to implement, or respond and not add anything if you disagree with the comment.

benbarsdell committed:
Remove unused header from meta_optimizer.cc

benbarsdell committed:
Fix TypeAttrId constructors

benbarsdell committed:
Add comment describing type_index

benbarsdell committed:
Remove unused arg and variable in GetTypeAttrs()

benbarsdell committed:
Expand docstring of GraphTypeTopologyView

benbarsdell committed:
Fix potential invalid iterator access

benbarsdell committed:
Remove unused variable in AllowedDataTypes

benbarsdell committed:
Rename NodeSet to NodeTypeIdSet

benbarsdell committed:
Fix log message in LogSkippedNode()

benbarsdell committed:
Make GetTailOfChain() return const

benbarsdell committed:
Fix comment in AddWhitelistOps

benbarsdell committed:
Remove unnecessary condition on Identity nodes

benbarsdell committed:
Remove unnecessary condition in DFS traversal

benbarsdell committed:
Add comment about use of IsIdentityAfterVariable

benbarsdell committed:
Handle fixed data types in AllowedDataTypes()

benbarsdell committed:
Fix docstring in _conv_pool

benbarsdell committed:
Use full names for protobuf modules

benbarsdell committed:
Enable Python tests to be run on pre-Volta GPUs

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -0,0 +1,1720 @@
+        type_index(_type_index),
+        fixed_type(DT_INVALID) {}
+
+  explicit TypeAttrId(DataType _fixed_type) : fixed_type(_fixed_type) {}
```
Fixed.

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -0,0 +1,1720 @@
+        const auto it2 = node_type_name_to_index_.find(
+            NodeTypeKey(input_node.name(), input_type_attr));
+        if (it2 == node_type_name_to_index_.end()) {
+          if (!skip_invalid_edges_) {
```
Fixed.

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -0,0 +1,1720 @@
+
+bool AutoMixedPrecisionImpl::IsOnGPU(const NodeDef& node) const {
+  string device_name;
+  if (node.device().empty()) {
```
I'm not sure I understand this one, could you clarify? The identical code is used in layout_optimizer.cc.

reedwm replied:
If layout_optimizer.cc does this as well, it is fine.

The point I was getting at was a user could do something like:

```
with tf.device('/replica:0'):
  x = <white-list op>
```

The optimizer would not touch `x`, despite it being on the GPU, because the string "GPU" does not appear in the device, and the device is not empty.

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -0,0 +1,1720 @@
+  for (int root_idx = 0; root_idx < graph_type_view_.num_nodes(); ++root_idx) {
+    const NodeTypeId& root = *graph_type_view_.GetNode(root_idx);
+    if (!ShouldProcess(*root.node) || downstream_of_white_set.count(root_idx) ||
+        !fp16_whitelist_.count(root.node->op())) {
```
Removed.

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -0,0 +1,1720 @@
+                  ShouldProcess(*item.node) && IsFloat32(item) &&
+                  SupportsFloat16(item) &&
+                  (fp16_clearlist_.count(item.node->op())) &&
+                  !IsIdentityAfterVariable(*item.node));
```
Done.

benbarsdell commented on tensorflow/python/grappler/auto_mixed_precision_test.py:
```diff
@@ -0,0 +1,555 @@
+    inp_b -> loop [ body_b ] -> out_b
+    """
+    if test.is_gpu_available(cuda_only=True,
+                             min_cuda_compute_capability=self.MIN_GPU_ARCH):
```
Done.

benbarsdell committed:
Add TODO regarding LookUpOpDef for functions

- Also handle failure inside SupportsFloat16().

benbarsdell committed:
Add TODO regarding handling of existing fp16 nodes

benbarsdell committed:
Add TODO regarding use of GraphTopologyView

benbarsdell committed:
Clarify/add comments describing blacklist pass

- Also removes unused white_set parameter.

benbarsdell committed:
Add TODO regarding tests for eager + tf.function

benbarsdell committed:
Replace intertwined loop tests with simple loops

- Intertwined loops are apparently not well tested in the rest of TF.
- The new simple loop tests capture most of the same color propagation
  behavior, and have simpler numerical behavior.

benbarsdell committed:
Use NaN-proof ops for tests, to avoid silent fails

benbarsdell committed:
Remove checks on no. casts in dropout test

- The dropout subgraph changes between TF versions, which affects the
  required number of casts.

benbarsdell committed:
Add note regarding fixed types in NodeTypeAttrMap

benbarsdell committed:
Move Sum to blacklist

- The dice loss function used in U-net uses Sums that overflow in fp16.
- This may reduce performance in other models if Sum is used in the bwd
  pass, but currently there is no way to identify such cases as being
  safe.
- Dropout is one such case where a Sum appears in the gradients, so the
  test needed to be adjusted.

benbarsdell committed:
Add OneHot to clearlist

benbarsdell committed:
Add contrib.rnn ops to the whitelist

- These all use CUBLAS on the GPU.

benbarsdell committed:
Return early if no whitelist ops found

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -0,0 +1,1720 @@
+  Status AddNode(const NodeDef& node) {
+    const OpDef* op_def_ptr = nullptr;
+    TF_RETURN_IF_ERROR(
+        OpRegistry::Global()->LookUpOpDef(node.op(), &op_def_ptr));
```
I've added a TODO (and also fixed the call to LookUpOpDef() inside SupportsFloat16()).

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -0,0 +1,1720 @@
+  //    "clearlist" ops).
+  //    This is done to prevent numerically-dangerous ops and their downstream
+  //    effects from being changed to fp16, which would risk breaking the
+  //    convergence of the model.
```
I've rewritten the description; let me know if it sounds OK now.

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -0,0 +1,1720 @@
+  }
+}
+
+void AutoMixedPrecisionImpl::PropagateBlackFwdThroughClearAndGray(
```
I've added a comment with the new description. I also removed the unused white_set argument.

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -0,0 +1,1720 @@
+  DataStructureOpsMap object_clients_map;
+  VLOG(2) << "Identifying Stack*V2 nodes";
+  TF_RETURN_IF_ERROR(AddDataStructureOpsToMap(
+      "StackV2", TypeAttrId("elem_type"), {{"StackPushV2", TypeAttrId("T")}},
```
1. Yes the same logic would need to be applied to Queue*V2 ops if they were added to the clearlist. Currently they are unlisted and so are never painted white.

2. These ops are on the clearlist, so they can be painted white without being between white nodes. In addition, ephemeral edges are added between writers and readers of the same data structure, so in this sense they do have both inputs and outputs.

benbarsdell commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -0,0 +1,1720 @@
+                 (!downstream_of_white_set.count(idx) &&
+                  !fp16_whitelist_.count(item.node->op()) &&
+                  !black_set.count(idx) && ShouldProcess(*item.node) &&
+                  IsFloat32(item) && SupportsFloat16(item) &&
```
I've added a TODO.

benbarsdell commented:
Thanks for the thorough review @reedwm. PTAL at my changes/comments in response.

There were a few very small things that we discovered after submitting the PR, so I've added those in too (see the last 4 commits).

reedwm commented on tensorflow/core/grappler/optimizers/auto_mixed_precision_lists.h:
```diff
@@ -0,0 +1,297 @@
+        "TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_CLEARLIST_REMOVE", "",
+        &to_remove);
+
+    auto list = std::set<string>{
```
I'm slightly worried that one may accidentally add a stateful op without realizing they also need to call AddDataStructureOpsToMap. Can you add a comment somewhere in this function, something like:

Note: if a stateful op is added, such as StackPopV2, is added to the clearlist, you must also modify AutoMixedPrecisionImpl to call AddDataStructureOpsToMap() with that op.

MattConley replied:
Added a comment to that effect

reedwm commented on tensorflow/python/grappler/auto_mixed_precision_test.py:
```diff
@@ -497,37 +511,30 @@ def testRecurrentLSTM(self):
-  def testPropagationThroughIntertwinedLoop7(self):
-    self._run_intertwined_loop_test('c', 'c', 'bWg', 'c', 'g', 'b', 2, 1)
+  def test_propagation_through_simple_loop_6(self):
+    self._run_simple_loop_test('b', 'CWCG', 'C')
```
Is the gray node made fp16 because it is "between" two whites nodes, where both white nodes are the same due to the loop? That sounds sensible, I am just checking.

MattConley replied:
That's correct; the black does not propagate into the loop because the clear nodes after the blacklisted node are not upstream of any blacklisted/greylisted nodes.  Then the traversal tags the grey node (and clear nodes) within the loop as both upstream and downstream of a whitelisted node (in this case the only one).

cbcase commented:
@reedwm @rmlarsen -- is there any special context we should have about the _ordering_ of the optimizers that run through MetaOptimizer?

I've been doing perf testing and noticed that we can run into unfortunate cases with `Sum` ops generated for the backward graph. As an example, consider a graph that runs `Dense+Relu -> Dropout -> Dense+Relu -> LossFn`. We'd like to run everything between the matmuls in fp16, both forward and backward.

But the (naive) backward graph for Dropout contains `Sum` ops (due to `BroadcastGradientArgs`). This breaks the span-finding behavior of AutoMixedPrecision, and it keeps dropout backward in fp32.

The constant folding optimizer correctly recognizes that these broadcasts aren't necessary and strips everything out. So if we run AutoMixedPrecision after constant folding, we get a better set of casts. However, we don't want to make AutoMixedPrecision a "run once at the end" pass (like xla-fusion), since there's some benefit to running things like constant folding _after_ -- for example it can push casts of constants to be constants of the casted type.

So, what I'd like to do is move AutoMixedPrecision toward the end of the (implicit) ordering in MetaOptimizer and keep it "run once," so that it can process an already-simplified graph but also benefit from downstream optimization.

Does that make sense? I'm anxious there is secret sauce in the existing ordering and don't want to break what's there.

rmlarsen commented:
@cbcase We normally run all grappler passes twice:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/meta_optimizer.cc#L369

so I think you can achieve the effect you desire by placing AutoMixedPrecision towards the end of the first iteration and  adding it to the set of run-once optimizers here:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/meta_optimizer.cc#L80

I agree that it would be nice to make the ordering more flexible and user defined.

MattConley committed:
Reorder auto mixed precision grappler pass

to take advantage of other optimizations

MattConley committed:
Merge del commands together

MattConley committed:
Remove MIN_GPU_ARCH check from tests.

MattConley committed:
Clarify comments describing propagation algorithm

MattConley committed:
Add comment about adding stateful ops to clearlist

MattConley commented on tensorflow/python/grappler/auto_mixed_precision_test.py:
```diff
@@ -151,81 +154,76 @@ def _lstm_cell(prev_c, prev_h, x):
-    return i<4
+    del c
+    del h
+    del ta_x
```
Sounds good, thanks

MattConley commented on tensorflow/python/grappler/auto_mixed_precision_test.py:
```diff
@@ -258,44 +256,57 @@ def _build_node_map(nodes):
 class AutoMixedPrecisionTest(test.TestCase):
   """Tests the Grappler auto mixed precision optimizer."""
-  MIN_GPU_ARCH = (7, 0)
+  MIN_GPU_ARCH = None
```
Updated as suggested

MattConley commented on tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:
```diff
@@ -1314,8 +1333,12 @@ void AutoMixedPrecisionImpl::AddWhitelistOps(
 
+// Adds nodes to black_set iff they are on the blacklist or they are on a
+// forward path from a blacklist node to a black/gray node (including the node
+// at the end of the path) through zero or more clear nodes.
```
Updated comments to clarify both grey and clear nodes are subject to black propagation

reedwm commented:
@rmlarsen, can you take a quick look at this PR now?

rthadur commented:
@MattConley can you please check failed builds

MattConley commented:
Looks like this change to VirtualPlacer's constructor is causing the break: https://github.com/tensorflow/tensorflow/commit/f089b3180ec7ddecd503d6d499cd6977c22b8f11

Working on a fix now

MattConley committed:
Merge remote-tracking branch 'origin/master' into mixed-precision-optimizer

MattConley committed:
Update VirtualPlacer call to match new api

MattConley commented:
After the VirtualPlacer update, the build looks good on my end

MattConley commented:
@reedwm @rmlarsen This should be ready for final review; let me know if I should change anything for the internal builds.

rmlarsen commented on tensorflow/core/grappler/optimizers/auto_mixed_precision_lists.h:
```diff
@@ -0,0 +1,300 @@
+        "TF_AUTO_MIXED_PRECISION_GRAPH_REWRITE_CLEARLIST_REMOVE", "",
+        &to_remove);
+
+    auto list = std::set<string>{
```
We usually use this pattern for large static string sets: (from grappler/op_types.cc)

static const gtl::FlatSet<string>* const kNonForwardingOps = CHECK_NOTNULL(
      (new gtl::FlatSet<string>{"ArgMax",
                                "ArgMin",
                                "AudioSpectrogram",
                                "BatchMatMul",
...

Same everywhere. std::set is very slow and can fragment memory.

MattConley replied:
Sounds good, thanks.  I replaced the std::set instances with gtl::FlatSet, let me know if I should change it.

rmlarsen commented on tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc:
```diff
@@ -0,0 +1,479 @@
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
```
In general, can you please use  EvaluateNodes(...) to run the original and modified graphs and verify that the outputs agree to the accuracy you expect.

Example:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/arithmetic_optimizer_test.cc#L3006
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/arithmetic_optimizer_test.cc#L3060

benbarsdell replied:
Would a TODO be acceptable for this PR? These tests were written in the style of loop_optimizer_test.cc and were not designed for numerical analysis. In addition to requiring a rewrite, adding numerical checks will also likely require careful changes to the test graphs to ensure they reflect real-world numerics.
Numerical correctness is also already checked in the Python tests.

rmlarsen replied:
I think simply making sure the graphs run after the rewrite would be useful. A TODO is fine for using more real-world numerics.

benbarsdell replied:
Done.
I threw in numerical checks too, with a TODO noting that they should be made more realistic.

rmlarsen replied:
Thanks!

MattConley committed:
Use gtl::FlatSet for op lists instead of std::set

benbarsdell committed:
Rewrite C++ tests to evaluate and verify output

rahulpalamuttam commented:
@MattConley The default behavior is to have mixed precision on by default?
I couldn't find a flag to toggle it on/off so just making sure.

benbarsdell commented:
@rahulpalamuttam no, it defaults to off:
https://github.com/tensorflow/tensorflow/pull/26342/files#diff-5b48a50cc60c18bb944e33e9c039f1c7R113
This toggle must be set to ON to enable it:
https://github.com/tensorflow/tensorflow/pull/26342/files#diff-280f21be5641238b9220b8d5df758555R88

asispatra commented:
To make use of Mixed precision, do I need to set any environment variable similarly what NVIDIA has `TF_ENABLE_AUTO_MIXED_PRECISION`?

OR 

Is the variable different here?

Thanks?

tfboyd commented:
@asispatra   This week we will have a full example added to 
https://github.com/tensorflow/models/tree/master/official/resnet

I am not sure how to alert you as to when it is finished.  We are adding it there to compare perf with the custom rolled solution and for continuous integration testing.  I am pretty sure you do not need the ENV VAR at all just using the auto loss scaling optimizer.  Feel free to ping me on this PR here WED this week or after assuming I have not gotten back to this thread to provide the link and more info.

reedwm commented:
The recommended way to use mixed precision is to wrap your optimizer with [enable_mixed_precision_graph_rewrite](https://github.com/tensorflow/tensorflow/blob/66d4f9e6d8f65c56eda1091384c4f0967abd0fdf/tensorflow/python/training/experimental/mixed_precision.py#L79):

```
optimizer = ...
optimizer = tf.train.experimental.enable_mixed_precision_graph_rewrite(optimizer)
```

As Toby said, there is no environmental variable in upstream TensorFlow, and we will update the official model with this change this week. 

Nvidia did publish a fork of TensorFlow in their [NDG container](https://devblogs.nvidia.com/nvidia-automatic-mixed-precision-tensorflow/), where mixed precision can be enabled with an environmental variable, but in upstream you instead wrap the optimizer.

asispatra commented:
@tfboyd and @reedwm, Thanks for the information. I will also keep watching this PR till WED to get more update.

For this PR, with there be any change in the [Tensorflow Benchmarks](https://github.com/tensorflow/benchmarks) code? I am asking this, because while I was trying NVIDIA's Tensorflow AutomaticMixedPrecision with Benchmark code, it was failing. 

Thanks.

tfboyd commented:
@asispatra we have partial stepped away from tensorflow/benchmarks for the moment.  Our focus is bringing the examples in tensorflow/models/official up to speed to benefit users of the high level APIs.  That said, I suspect @reedwm would address issues you are having and it would be nice to have "AMP" in the tensorflow/benchmark.  

The long story shortened is we are using tf_cnn_benchmark as our high level benchmark for ResNet but beyond that we are not doing much with it currently.  But we do not want it broken for TF 1.x

reedwm commented:
I didn't plan on adding support to tf_cnn_benchmarks, but I might change my mind. One issue is that tf_cnn_benchmarks uses tf.gradients to compute gradients, while `enable_mixed_precision_graph_rewrite` requires you to call optimizer.compute_gradients.

Note tf_cnn_benchmarks already supports fp16 with manual casts. This is very ugly and difficult compared to `enable_mixed_precision_graph_rewrite`, but the rest of tf_cnn_benchmarks is fairly messy as well.

rahulpalamuttam commented:
@reedwm 
Can we use mixed precision without loss_scaling i.e. a value of None instead of 'dynamic'?
Or could we just give it a value of 1.0?

reedwm commented:
You should keep loss_scale as 'dynamic'. Giving it 1.0 is equivalent to not doing loss scaling, but without loss scaling, your model may converge to a worse accuracy. You cannot pass None.

rllin-fathom commented:
has anybody tested this with horovod?

I would be replacing this line basically to wrap a horovod optimizer

https://github.com/NVIDIA/DeepLearningExamples/blob/master/TensorFlow/LanguageModeling/BERT/optimization.py#L80

tfboyd commented:
No but Nvidia did the PR so I am pretty sure it works.  They have a Bert
with FP16 and XLA that is very fast in a blog.  It may predate this exact
code path.

On Fri, May 24, 2019, 12:21 PM Randall Lin <notifications@github.com> wrote:

> has anybody tested this with horovod?
>
> I would be replacing this line basically to wrap a horovod optimizer
>
>
> https://github.com/NVIDIA/DeepLearningExamples/blob/master/TensorFlow/LanguageModeling/BERT/optimization.py#L80
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/26342?email_source=notifications&email_token=AFTF5MWTAYOCGK7MLDGHQU3PXA5Z7A5CNFSM4G3W4WM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWGK62I#issuecomment-495759209>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFTF5MTLRMW2EZ6DPS4OHDLPXA5Z7ANCNFSM4G3W4WMQ>
> .
>

rllin-fathom commented:
@tfboyd yea I'm basically trying to replicate their squad benchmarks right now but on an image with tf nightly + horovod based on https://gitlab.com/nvidia/cuda/tree/ubuntu18.04 (vs using the NGC image which is a blackbox)

rllin-fathom commented:
I have reproduced the original Nvidia benchmark on SQUAD on 1x V100 16Gi using tf nightly based off this (open) image https://gitlab.com/nvidia/cuda (the original benchmark uses https://docs.nvidia.com/deeplearning/frameworks/tensorflow-release-notes/rel_19.03.html#rel_19.03 which is closed source)

See https://github.com/NVIDIA/DeepLearningExamples/issues/53#issuecomment-495808640 for details.

However I run into an OOM issue with 8x V100 16Gi.  More info here https://github.com/NVIDIA/DeepLearningExamples/issues/53#issuecomment-495808640

Any insights or assistance would be appreciated, as I haven't found an open source reproduction of multi-gpu with fp16 (AMP) with xla on horovod.

benbarsdell commented:
@rllin-fathom it's possible that we're using different definitions of closed source, but as I mentioned in the other thread, the NGC images do contain full source code for the TensorFlow build that they include, plus instructions for how to rebuild with your own customizations.

rllin-fathom commented:
@benbarsdell mm yes you're right it does, I was speaking towards the fact that it's unclear if the image has any other customizations not apparent in the release notes.  I've (for now, based on confirmation from you and @nluehr (thanks!)) assumed that the tf packaged in the image is functionally identical to tf nightly (both have AMP).

Thus the remaining unknown for me is if there are image build customizations, that I am not cloning.  But I did try to follow the release notes at https://docs.nvidia.com/deeplearning/frameworks/tensorflow-release-notes/rel_19.03.html#rel_19.03 and horovod installation instructions at https://github.com/horovod/horovod/blob/master/docs/gpus.md  and https://github.com/horovod/horovod/blob/master/Dockerfile.test.gpu

rllin-fathom commented:
e.g.

which any flags during horovod installation were turned on https://github.com/horovod/horovod/blob/master/docs/gpus.md#advanced-have-a-proprietary-mpi-implementation-with-gpu-support-optimized-for-your-network

xpirad commented:
How do i enable auto mixed precision in keras with tensorflow?

reedwm commented:
@xpirad, you can use auto mixed precision by wrapping your optimizer as follows

```
opt = ... # A TensorFlow optimizer
opt = tf.train.experimental.enable_mixed_precision_graph_rewrite(opt)
```

This works regardless of whether you use tf.keras or not. See the [enable_mixed_precision_graph_rewrite docstring](https://github.com/tensorflow/tensorflow/blob/165a535d07750de0d12747649423259107ec6539/tensorflow/python/training/experimental/mixed_precision.py#L75) for details.

We are also working on a pure tf.keras API for TF 2, but this API is not yet ready.

yselivonchyk commented:
@reedwm what is the lowest supported Tensorflow version?

reedwm commented:
@yselivonchyk, this is available in TF 1.14 or higher.

sgambient commented:
Is Mixed Precision only  supported for optimization in TF 1.14 ?  Is there a way to enable Mixed Precision  for inference ?

nluehr commented:
You can enable the graph rewrite pass by setting the following session config option.

`config.graph_options.rewrite_options.rewrite_options.auto_mixed_precision = 1`

sgambient commented:
Thanks. So the NVIDIA recommended way of setting  TF_ENABLE_AUTO_MIXED_PRECISION=1 in the environment does not apply ?   Also, Mixed Precision has to be specifically enabled , not auto-detected ?

nluehr commented:
TF_ENABLE_AUTO_MIXED_PRECISION only applies to the NVIDIA NGC TensorFlow Docker container. And yes, enabling mixed precision must be enabled by setting the config option above (to enable only the graph rewrite pass) or calling enable_mixed_precision_graph_rewrite() to enable both the graph rewrite and wrap the optimizer with loss scaling.

x10000year commented:
Will we have TF_ENABLE_NHWC soon for official tensorflow? TF_ENABLE_NHWC is important to help tftrt optimize graph written in NHWC layout, otherwise there will be a lot of conversion between NHWC/NCHW generated by tftrt on volta/turing gpus. Since NHWC layout has the best performance for tensor core, tftrt should support graphes in NHWC layout.

I also observed that if the graph is written in NCHW layout, tftrt still generates a lot of alternating NCHW/NHWC conversions on volta gpus. Such conversions are redundant and should be automatically removed.

tfboyd commented:
As of tf-nightly-gpu and f-nightly-gpu-2.0-preview 2.0.0-dev20190723  there is an automatic pass with what is known as "grappler" that will rewrite NCHW to NHWC if you are using FP16.  The only model I am testing for the scenario is ResNet50 V1.5 and perf with real data improved from 767 to 1,008 image/sec without using XLA.  The tests are done with the data format in the code of channels first (NCHW); and "grappler" does the change.  It should work for any model; but I want to be clear I am only testing the one.  

Looking toward the foggy future.  The XLA number is 1,313; but multi-GPU (even single node) is not well supported.  FYI if you were using FP16 with XLA before it would have already been flipping to NHWC and it has a few more tricks.  Again, not ready for production

This was all done using the high level APIs with these models.

- [Keras](https://github.com/tensorflow/models/tree/master/official/resnet/keras)
- [Old Skool Estimator and tf.layers](https://github.com/tensorflow/models/tree/master/official/r1/resnet)

x10000year commented:
@tfboyd Actually my model was manually coded in NHWC and FP16 at the first time, but tftrt didn't optimize well for this model since it generated too many redundant conversions between NHWC/NCHW. Then I manually modify the model to use NCHW, and tftrt works better now, but the generated tft graph still have many redundant conversions. All tests are done on turing gpus. So I think a grappler that rewrites NCHW to NHWC doesn't help much, since this could be easily done manually.

don-tpanic commented:
> @xpirad, you can use auto mixed precision by wrapping your optimizer as follows
> 
> ```
> opt = ... # A TensorFlow optimizer
> opt = tf.train.experimental.enable_mixed_precision_graph_rewrite(opt)
> ```
> 
> This works regardless of whether you use tf.keras or not. See the [enable_mixed_precision_graph_rewrite docstring](https://github.com/tensorflow/tensorflow/blob/165a535d07750de0d12747649423259107ec6539/tensorflow/python/training/experimental/mixed_precision.py#L75) for details.
> 
> We are also working on a pure tf.keras API for TF 2, but this API is not yet ready.

Thanks for the solution. Not sure if I'm doing it wrong, but I tried it and got:

```
WARNING:tensorflow:You already have existing Sessions that do not use mixed precision. enable_mixed_precision_graph_rewrite() will not affect these Sessions.
```

reedwm commented:
@don-tpanic, try calling `enable_mixed_precision_graph_rewrite` before creating the VGG16 model.

lzha106 commented:
> You can enable the graph rewrite pass by setting the following session config option.
> 
> `config.graph_options.rewrite_options.rewrite_options.auto_mixed_precision = 1`

two "rewrite_options" can't work, correct it to be:
`config.graph_options.rewrite_options.auto_mixed_precision = 1`
